### ISSUE TITLE
split function missing

### ISSUE THREAD

**Issue by priyankat99** [fancy-regex/fancy-regex#104]
i wanted to split a string with regex and was wondering if theres a reason that that function is missing? its implemented in the rust-regex crate and also their regex implements pattern so the std::split is an option as well. 
i am interested in adding it but wondering if it was purposely left out for some reason?

**Comment by keith-hall** [comment]
I believe it has only been left out because nobody has needed it yet and thus nobody has implemented it yet. There have been some issues reported in the past where people asked for `fancy-regex` to implement some APIs that `regex` has and to my knowledge they have never been rejected.

**Comment by juan-coralogix** [comment]
It would be an awesome feature.

**Comment by fdncred** [comment]
definitely needed

**Comment by wez** [comment]
I'm also missing this function.  I whipped up the following to satisfy my need, thought I'd share here in case someone has the bandwidth to make a proper PR out of it.
The logic here is a interpretation of https://github.com/rust-lang/regex/blob/8856fe36ac7dc37989e6ffb26b5fc57189bae626/regex-automata/src/meta/regex.rs#L2180-L2219 but adapted to work on the public types available from this crate.

```rust

struct Split<'r, 'h> {
    finder: Matches<'r, 'h>,
    last: usize,
    haystack: &'h str,
}

impl<'r, 'h> Split<'r, 'h> {
    fn split(re: &'r Regex, haystack: &'h str) -> Self {
        Self {
            finder: re.find_iter(haystack),
            last: 0,
            haystack,
        }
    }
}

impl<'r, 'h> Iterator for Split<'r, 'h> {
    type Item = Result<&'h str, fancy_regex::Error>;

    fn next(&mut self) -> Option<Result<&'h str, fancy_regex::Error>> {
        match self.finder.next() {
            None => {
                let len = self.haystack.len();
                if self.last > len {
                    None
                } else {
                    let span = &self.haystack[self.last..len];
                    self.last = len + 1; // Next call will return None
                    Some(Ok(span))
                }
            }
            Some(Ok(m)) => {
                let span = &self.haystack[self.last..m.start()];
                self.last = m.end();
                Some(Ok(span))
            }
            Some(Err(e)) => Some(Err(e)),
        }
    }
}

fn split_into_vec(re: &Regex, haystack: &str) -> Result<Vec<String>, fancy_regex::Error> {
    let mut result = vec![];
    for m in Split::split(re, haystack) {
        let m = m?;
        result.push(m.to_string());
    }
    Ok(result)
}

#[cfg(test)]
mod test {
    use super::*;

    #[test]
    fn fancy_split() {
        let re = Regex::new("[ \t]+").unwrap();
        let hay = "a b \t  c\td    e";
        let fields = split_into_vec(&re, hay).unwrap();
        assert_eq!(fields, vec!["a", "b", "c", "d", "e"]);
    }
}

```

**Comment by k94-ishi** [comment]
@wez 
I was able to create a pull request using your code as a reference. Thank you!

**Comment by robinst** [comment]
`split` and `splitn` have been released in 0.14.0, thanks @k94-ishi 🎉 ! https://github.com/fancy-regex/fancy-regex/blob/main/CHANGELOG.md#0140---2024-10-24
